mlfinlab/labeling/labeling.py

```diff
@@ -0,0 +1,288 @@
+
+    This function applies the triple-barrier labeling method. It works on a set of
+    datetime index values (molecule). This allows the program to parallelize the processing.
+
+    :param close: (series) close prices
+    :param events: (series) of indices that signify "events" (see get_t_events function
+    for more details)
+    :param pt_sl: (array) element 0, indicates the profit taking level; element 1 is stop loss level
+    :param molecule: (an array) a set of datetime index values for processing
+    :return: Dataframe of timestamps of when first barrier was touched
+    """
+    # Apply stop loss/profit taking, if it takes place before t1 (end of event)
+    events_ = events.loc[molecule]
+    out = events_[['t1']].copy(deep=True)
+
+    profit_taking_multiple = pt_sl[0]
+    stop_loss_multiple = pt_sl[1]
+
+    # Profit taking active
+    if profit_taking_multiple > 0:
+        profit_taking = profit_taking_multiple * events_['trgt']
+    else:
+        profit_taking = pd.Series(index=events.index)  # NaNs
+
+    # Stop loss active
+    if stop_loss_multiple > 0:
+        stop_loss = -stop_loss_multiple * events_['trgt']
+    else:
+        stop_loss = pd.Series(index=events.index)  # NaNs
+
+    # Get events
+    for loc, vertical_barrier in events_['t1'].fillna(close.index[-1]).iteritems():
+        df0 = close[loc: vertical_barrier]  # path prices
+        df0 = (df0 / close[loc] - 1) * events_.at[loc, 'side']  # path returns
+        out.loc[loc, 'sl'] = df0[df0 < stop_loss[loc]].index.min()  # earliest stop loss
+        out.loc[loc, 'pt'] = df0[df0 > profit_taking[loc]].index.min()  # earliest profit taking
+
+    return out
+
+
+# Snippet 3.4 page 49, Adding a Vertical Barrier
+def add_vertical_barrier(t_events, close, num_days=1):
+    """
+    Snippet 3.4 page 49, Adding a Vertical Barrier
+
+    For each index in t_events, it finds the timestamp of the next price bar at or immediately after
+    a number of days num_days. This vertical barrier can be passed as an optional argument t1 in get_events.
+
+    This function creates a series that has all the timestamps of when the vertical barrier would be reached.
+
+    :param t_events: (series) series of events (symmetric CUSUM filter)
+    :param close: (series) close prices
+    :param num_days: (int) maximum number of days a trade can be active
+    :return: (series) timestamps of vertical barriers
+    """
+
+    vertical_barriers = close.index.searchsorted(t_events + pd.Timedelta(days=num_days))
+    vertical_barriers = vertical_barriers[vertical_barriers < close.shape[0]]
+    vertical_barriers = pd.Series(close.index[vertical_barriers],
+                                  index=t_events[:vertical_barriers.shape[0]])  # NaNs at end
+    return vertical_barriers
+
+
+# Snippet 3.3 -> 3.6 page 50, Getting the Time of the First Touch, with Meta Labels
+def get_events(close, t_events, pt_sl, target, min_ret, num_threads, vertical_barrier_times=False, side=None):
```

Comment:
Can we also rename function so that from naming user understands, that it is meta_labeling + TripleBarrier?

Reply:
I see this as the main triple barrier method, apply_pt_sl_on_t1 is the time of first touch. What do you think?